internal/langserver/handlers/execute_command_module_providers_test.go

```diff
@@ -0,0 +1,168 @@
+		Method: "workspace/executeCommand",
+		ReqParams: fmt.Sprintf(`{
+		"command": %q
+	}`, cmd.Name("module.providers"))}, code.InvalidParams.Err())
+}
+
+func TestLangServer_workspaceExecuteCommand_moduleProviders_basic(t *testing.T) {
+	modDir := t.TempDir()
+	modUri := uri.FromPath(modDir)
+
+	s, err := state.NewStateStore()
+	if err != nil {
+		t.Fatal(err)
+	}
+
+	err = s.Modules.Add(modDir)
+	if err != nil {
+		t.Fatal(err)
+	}
+
+	metadata := &tfmod.Meta{
+		Path:             modDir,
+		CoreRequirements: testConstraint(t, "~> 0.15"),
+		ProviderRequirements: map[tfaddr.Provider]version.Constraints{
+			tfaddr.NewDefaultProvider("aws"):    testConstraint(t, "1.2.3"),
+			tfaddr.NewDefaultProvider("google"): testConstraint(t, ">= 2.0.0"),
+		},
+		ProviderReferences: map[tfmod.ProviderRef]tfaddr.Provider{
+			{LocalName: "aws"}:    tfaddr.NewDefaultProvider("aws"),
+			{LocalName: "google"}: tfaddr.NewDefaultProvider("google"),
+		},
+	}
+
+	err = s.Modules.UpdateMetadata(modDir, metadata, nil)
+	if err != nil {
+		t.Fatal(err)
+	}
+
+	pVersions := map[tfaddr.Provider]*version.Version{
+		tfaddr.NewDefaultProvider("aws"):    testVersion(t, "1.2.3"),
+		tfaddr.NewDefaultProvider("google"): testVersion(t, "2.5.5"),
+	}
+	err = s.Modules.UpdateInstalledProviders(modDir, pVersions)
+	if err != nil {
+		t.Fatal(err)
+	}
+
+	ls := langserver.NewLangServerMock(t, NewMockSession(&MockSessionInput{
+		TerraformCalls: &exec.TerraformMockCalls{
+			PerWorkDir: map[string][]*mock.Call{
+				modDir: validTfMockCalls(),
+			},
+		},
+		StateStore: s,
+	}))
+	stop := ls.Start(t)
+	defer stop()
+
+	ls.Call(t, &langserver.CallRequest{
+		Method: "initialize",
+		ReqParams: fmt.Sprintf(`{
+	    "capabilities": {},
+	    "rootUri": %q,
+		"processId": 12345
+	}`, modUri)})
+	ls.Notify(t, &langserver.CallRequest{
+		Method:    "initialized",
+		ReqParams: "{}",
+	})
+
+	ls.CallAndExpectResponse(t, &langserver.CallRequest{
+		Method: "workspace/executeCommand",
+		ReqParams: fmt.Sprintf(`{
+		"command": %q,
+		"arguments": ["uri=%s"]
+	}`, cmd.Name("module.providers"), modUri)}, `{
+		"jsonrpc": "2.0",
+		"id": 2,
+		"result": {
+			"v": 0,
+			"provider_requirements": {
+				"registry.terraform.io/hashicorp/aws": {
+					"display_name": "hashicorp/aws",
+					"version_constraint":"1.2.3"
+				},
+				"registry.terraform.io/hashicorp/google": {
+					"display_name": "hashicorp/google",
+					"version_constraint": "\u003e= 2.0.0"
+				}
+			},
+			"installed_providers":{
+				"registry.terraform.io/hashicorp/aws": "1.2.3",
+				"registry.terraform.io/hashicorp/google": "2.5.5"
+			}
+		}
+	}`)
+}
+
+func testConstraint(t *testing.T, v string) version.Constraints {
+	constraints, err := version.NewConstraint(v)
+	if err != nil {
+		t.Fatal(err)
+	}
+	return constraints
+}
+
+func testVersion(t *testing.T, v string) *version.Version {
+	ver, err := version.NewVersion(v)
+	if err != nil {
+		t.Fatal(err)
+	}
+	return ver
+}
```

Comment:
Nitpick - you could probably replace this with the upstream `version.Must()` which will `panic` on error and the panic would show up in test (should the version be unparseable), maybe not as nicely, but it will be pretty obvious where it comes from. The main benefit is that it's less code to maintain here 🤷🏻‍♂️  😄